docs/troubleshooting.md

```diff
@@ -1,79 +1,165 @@
-For vscode you can read how to access the lsp session log [here])https://github.com/haskell/vscode-haskell#investigating-and-reporting-problems).
+
+## Common issues
+
+### Wrong server binary being used
+
+HLS needs to be compiled against the same version of GHC as is used in the project.
+Normally, we ship binaries for multiple versions and `haskell-language-server-wrapper` selects the correct one.
+
+If you see an error about HLS being compiled with the wrong version of GHC, then you either need to install the correct one (if you installed it yourself), or there is something going wrong with the wrapper selecting the right HLS binary to launch.
+
+### Unsupported GHC version
+
+HLS does not support every GHC version - there are a lot of them!
+Please see the [supported versions page](./supported-versions.md) for more information.
+
+In particular, support for GHC 9.0 and 9.2 is only partial; some features are unavailable with GHC >= 9.0.1, and HLS is not available at all for GHC 9.2.
+
+### Missing server or build tools
 
 The most common client-related problem is the client simply not finding the server executable or the tools needed to load Haskell code (`ghc`, `cabal`, or `stack`). So make sure that you have the right `PATH` and you have configured the client to look for the right executables.
+
+Usually this will be visible in the client's log.
+
+### Compilation failures
+
+Sometimes HLS will simply fail to do anything with a file, or give nonsensical error messages.
+The most common cause of this is that HLS is using the wrong `hie-bios` cradle to decide how to build the project (e.g., trying to use `stack` instead of `cabal`).
+The server log will show which cradle is being chosen.
+
+Using an explicit `hie.yaml` to configure the cradle can resolve the problem, see the [configuration page](./configuration.md#configuring-your-project-build).
+
+### Problems with dynamic linking
+
+The prebuilt binaries which we provide are statically linked, which means that they don't work well with projects which use dynamic linking.
+A typical symptom is the presence of errors containing `unknown symbol`, and it is common for users Arch Linux, which ships a dynamically linked version of `ghc`.
+
+The workaround is to use a version of HLS compiled from source with the `ghc` option `-dynamic` enabled. See more details [here](https://github.com/haskell/haskell-language-server/issues/1160#issuecomment-756566273).
+
+### Support for Template Haskell
+
+Template Haskell should work fine in Linux and Windows with the distributed binaries.
+In MacOs a dynamically linked binary of HLS is required, otherwise use of TH can lead to segmentation faults.
+The easiest way to obtain a dynamically linked HLS binary is to build it locally.
+With `cabal` this can be done as follows:
+
+    cabal update && cabal install haskell-language-server --enable-executable-dynamic
```

Comment:
```suggestion

### Template Haskell

**tl;dr** Use a dynamically linked HLS binary.

The easiest way to obtain a dynamically linked HLS binary is to build it locally. With `cabal` this can be done as follows:

    cabal update && cabal install haskell-language-server --enable-executable-dynamic

With `stack` you need to manually add the ghc option `-dynamic`.

Static binaries use the GHC linker for dynamically loading dependencies when typechecking TH code, and this can run into issues when loading shared objects linked against mismatching system libraries, or into GHC linker  bugs (mainly the Mach linker used in Mac OS, but also potentially the ELF linker). Dynamically linked binaries (including`ghci`) use the system linker instead of the GHC linker and avoid both issues. 

Note: HLS binaries prior to 1.6.0 were statically linking `glibc` which is not a supported configuration and has been replaced by `musl`.

Reply:
Thank you!